File: 2401-2500/2401-2410/2405M_Optimal_Partition_of_String/bit.js
// 2405. Optimal Partition of String
// https://leetcode.com/problems/optimal-partition-of-string/
// T.C.: O(n)
// S.C.: O(1)
/**
 * @param {string} s
 * @return {number}
 */
var partitionString = function (s) {
  const a = 'a'.charCodeAt(0);
  let count = 1;
  let mask = 0;
  for (let i = 0; i < s.length; i++) {
    const ch = s.charCodeAt(i) - a + 1;
    if (mask & (1 << ch)) {
      count++;
      mask = 0;
    }
    mask |= 1 << ch;
  }
  return count;
};

var s = 'abacaba';
var expected = 4;
var result = partitionString(s);
console.log(result, result === expected);

var s = 'ssssss';
var expected = 6;
var result = partitionString(s);
console.log(result, result === expected);

File: 2401-2500/2401-2410/2405M_Optimal_Partition_of_String/kt/bit.kt
// 2405. Optimal Partition of String
// https://leetcode.com/problems/optimal-partition-of-string/
// T.C.: O(n)
// S.C.: O(1)
class Solution {
  fun partitionString(s: String): Int {
    var count = 1
    var mask = 0
    for (i in s.indices) {
      val ch = s[i] - 'a' + 1
      if (mask and (1 shl ch) != 0) {
        count++
        mask = 0
      }
      mask = mask or (1 shl ch)
    }
    return count
  }
}

fun main() {
  val solution = Solution()

  var s = "abacaba"
  var expected = 4
  var result = solution.partitionString(s)
  println("${result} -> ${result == expected}")

  s = "ssssss"
  expected = 6
  result = solution.partitionString(s)
  println("${result} -> ${result == expected}")
}
